sktime/forecasting/callbacks/mlflow.py

```diff
@@ -0,0 +1,242 @@
+"""MLFLow callback for logging metrics and plots to MLFlow."""
+from typing import Optional
+
+import mlflow
```

Comment:
This is the main culprit I believe. It's not a core dependency, and even has a soft dependency has so many other requirements. This import should go inside a function call, similar to how we did in statsforecast.

Reply:
Then i have to import mlflow 3 times though

Reply:
whats the point of importing things inside a function? We will have to reimport again and again

Reply:
If we do not import internally in a function, then it becomes a mandatory dependency. For example in this case, every user of sktime (of forecasting comonent or anything else) has to install mlflow (and all the dependencies that it implies) even though they do not use it themselves.

Reply:
I see, it is usually not recommended to do that though. Have you seen any other library or if instead of putting the import inside a function, changing something else in the library would be preferable? I think i have almost never seen this imports inside a function ever in a library. Any examples of well known libraries where they use this pattern?

Reply:
If you are comparing `sktime` with libraries like `numpy` or `scikit-learn`, then in my opinion they are not comparable. They implement the logic in the package python modules themselves and have very very few dependencies. They do not have worry about some 50-60 soft dependencies as it's there in `sktime` where quite possibly few are not even compatible to each other and separation is necessary.

The comparison should be made with other framework libraries, for example the recently super popular `langchain` or `llama-index`. Every new libraries or etc. wants an integration through them, but you don't install them all and only what you need. For example see non-global imports here for [Chroma in Langchain](https://github.com/langchain-ai/langchain/blob/de496062b3e740aed2c7097424749d1145e4aaab/libs/community/langchain_community/vectorstores/chroma.py#L80-L87).

Or if you consider `ray`, suppose you want to use `BOHB`. You do not get that support just with installing `ray[tune]`, you specifically need to install the dependency separately so that other uses are not affected. Or, for example Excel suport in `pandas`, where you install `openpyxl` or similar only when you need and not always, which is protected by `import_optional_dependency` function.

Let me know if these examples are insufficient. At top level developer code, for example when I'm using `sktime` for my office work, I agree to your points and will stick to only global imports where I know I'll always use this as a dependency and it's guaranteed to be installed. But at the framework code level I don't think the same guidelines apply, or at least not as a best practice. These are just my thoughts/opinions of course, @fkiraly / @benHeid / @achieveordie can share their views.

Reply:
@yarnabrina  Changed the imports!

Reply:
Thanks, keeping this conversation open for now in case other core developers suggest different options.

Reply:
Also, just so you know, there are few other files where soft dependencies are checked just once at top of file and then imports are made. This leads to package imports even though they are not necessarily used (based on my understanding) and in theory can have some side effects (loading time or etc.). I do not like this approach personally, but these are present (in logic and in test) and being actively developed in new PR's as well, so it seems fair that you should know given you raised the point of multiple imports previously.

Reply:
@yarnabrina, yes, I'm aware of those imports and I'm actually trying to actively get rid of these due to loading time issues. If you see any, issue would be appreciated (or a PR that moves them deeper).

Reply:
> These are just my thoughts/opinions of course

Great explanation, I also learnt a few things.

Just to point this out, `sktime` was using the pattern before langchain 😁